beluga/include/beluga/sensor/data/occupancy_grid.hpp

```diff
@@ -0,0 +1,177 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_SENSOR_DATA_OCCUPANCY_GRID_HPP
+#define BELUGA_SENSOR_DATA_OCCUPANCY_GRID_HPP
+
+#include <cstdint>
+#include <tuple>
+#include <vector>
+
+#include <beluga/sensor/data/linear_grid.hpp>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <Eigen/Core>
+
+namespace beluga {
+
+/**
+ * \page OccupancyGrid2dPage Beluga named requirements: OccupancyGrid2d
+ *
+ * Occupancy grids model obstacle probability. These grids also define
+ * their own frame in the grid embedding space, adding so called global
+ * coordinates to regular (aka local) coordinates. Occupancy grids as
+ * defined in Beluga are linear grids.
+ *
+ * A type `G` satisfies `OccupancyGrid2d` requirements if it satisfies
+ * \ref LinearGrid2Page and given `g` a possible const instance of `G`:
+ * - `g.value_traits()` returns a value `t` of type `T` such that given a grid cell
+ *    data value `v`:
+ *    - `t.is_free(v)` returns true if the value is consistent with a free grid cell;
+ *    - `t.is_occupied(v)` returns true if the value is consistent with an occupied grid cell;
+ * - given possibly const grid cell index `i` of `std::size_t` type,
+ *   `g.free_at(i)` returns true if such cell is free;
+ * - given possibly const grid cell coordinates `xi` and `yi` of type `int`,
+ *   `g.free_at(xi, yi)` returns true if such cell is free;
+ * - given possibly const grid cell coordinates `pi` of `Eigen::Vector2i` type,
+ *   `g.free_at(p)` returns true if such cell is free;
+ * - given possibly const embedding space coordinates `x` and `y` of type `double`,
+ *   `g.free_near(x, y)` returns true if the nearest cell is free;
+ * - given possibly const embedding space coordinates `p` of `Eigen::Vector2d` type,
+ *   `g.free_near(p)` returns true if the nearest cell is free;
+ * - given possibly const grid cell index `i` of `std::size_t` type and frame `f`,
+ *   `g.coordinates_at(i, f)` returns embedding space coordinates in the corresponding
+ *   frame as an `Eigen::Vector2d` value;
+ * - given a possibly const range `r` of grid cell coordinates or indices and frame `f`,
+ *   `g.coordinates_for(r, f)` returns a range of embedding space coordinates in the
+ *   corresponding frame as `Eigen::Vector2d` values;
+ * - `g.free_cells()` returns a range of `std::size_t` indices to free grid cells;
+ * - `g.obstacle_data()` returns a range of `bool` values, representing grid cell occupancy;
```

Comment:
I think we're missing `origin()` here.

Reply:
Indeed. Fixed in https://github.com/Ekumen-OS/beluga/pull/188/commits/771ac50e97960cab31f16450833a2b23f58167fc.